### DISCUSSION TITLE
How to clear trial email?

### DISCUSSION THREAD

**Discussion by ba32107** [Glench/ExtPay#81] [Q&A] [answered]
Hi! My context is the following: I am just exploring ExtPay and I tried the "free trial" option with my developer email address. It worked great, next, I wanted to test if I can do another trial after reinstall. As expected, I wasn't able to, since ExtPay knows this email has already been used.

But - how do I reset this? There is no button to delete a free trial address on the portal. I will want to test this free trial flow a couple of times during development, and I don't really want to create different email addresses for this purpose. Can I delete a free trial email somehow?

**Comment by Glench** [comment, marked as answer]
I just fixed this so that devs using their dev email in dev mode extensions can have unlimited free trials.

**Reply by FlashInThePan** [reply]
So how is this done? I uninstalled the extension and reinstalled. That didn't put my dev email in free trial. Whats the process, please?

**Reply by Glench** [reply]
Just input your dev email exactly as it appears on your logged in dev account.

**Reply by FlashInThePan** [reply]
When I reinstall my extension subscription status is not being set to unpaid.

These are the steps I take:

1. I reinstall my extension.
2. I click on my extension to load the popup.
	Here I see that my subscription status is cancelled for some reason (not unpaid)
3. I open the start trial page and submit, click on the email confirmation and open my popup again.
4. In my console I see I have the message "service worker got message! Here it is: trial-start", however, subscription status is "cancelled". 

If the subscriptionstatus = "cancelled", I don't know to offer the user a free trial because it looks like the user has paid in the past so shouldn't get a trial. Similarly, beyond 7 days the cancelled status suggests that the user has already paid in the past so I don't know to tell the user their trial has expired.

I'm confused because unless, as you said, the user is actually set to unpaid when the extension is reinstalled, the workflow doesn't fit and I'm unable to test my trial and post trial functionality. 

Any advice would be appreciated.

**Reply by Glench** [reply]
I'm not really sure what you're doing. If you reinstall the extension then there is no subscription status set yet, so I don't know how you would be seeing "cancelled".

**Reply by FlashInThePan** [reply]
Like I said, I simply remove the extension from the extension manager and reinstall the extension using "unpack". The only direct interact with extpay is in the whole extension is right at the beginning of popup.js in order to write the user object to the console, as below:

![image](https://user-images.githubusercontent.com/98082973/226467276-5d62af96-a8b8-47c7-a3ca-9a61a3e23f04.png)

I simply click on the extension icon to bring up the popup, then inspect the popup page to see the user object:

![image](https://user-images.githubusercontent.com/98082973/226466025-c385eb5c-7710-4fc4-a988-7de000eb9dd3.png)

I don't interact with any payment or other forms before I see the Cancelled status for the user objects whatsoever.

Tbh mate, I'm really desperate to get this working coz my company has a roll out planned for end of this week, so any help would be appreciated - as I can't see what I can be doing wrong. I am doing exactly as I describe, so I don't know where to go from here.

**Reply by Glench** [reply]
Don't worry, we'll figure it out. There is no such status as capital C Cancelled so I'm really not sure how it got that way. There are no entries at all in the database with the status "Cancelled" which makes me think it's a local thing to your extension. Try deleting your extension's storage with `chrome.storage.sync.clear()` and removing and reinstalling your extension. If the same behavior still happens, please copy paste the contents of `console.log(await chrome.storage.sync.get(['extensionpay_api_key', 'extensionpay_user']))`.

**Reply by FlashInThePan** [reply]
Thanks so much for the support. Running chrome.storage.sync.clear() and removing and reinstalling my extension made no difference. The output of console.log(await chrome.storage.sync.get(['extensionpay_api_key', 'extensionpay_user'])) is:

![image](https://user-images.githubusercontent.com/98082973/226476585-823f6e09-8823-4fd7-bc7b-207ee9f72f88.png)

The user objects is the same as before.

**Reply by FlashInThePan** [reply]
Maybe coincidentally, but I see Cancelled with a capital C in the subscription maintenance form:
![image](https://user-images.githubusercontent.com/98082973/226477527-46038c1b-ae11-4e5d-8e93-f6116eca0bfd.png)

However, the value of the option in the select control in the form is all lowercase:

![image](https://user-images.githubusercontent.com/98082973/226477891-8fb752bc-b6d1-4336-b404-92cefb0c499d.png)

**Reply by Glench** [reply]
> Thanks so much for the support. Running chrome.storage.sync.clear() and removing and reinstalling my extension made no difference.

This is literally impossible as far as I know. I expect you're removing and reinstalling the wrong extension. All the local extensionpay user data is removed when an extension is removed.

> Maybe coincidentally, but I see Cancelled with a capital C in the subscription maintenance form:

That's just a cosmetic thing. And as you can see it has only 1 "L" not 2 like in your screenshot.

**Reply by FlashInThePan** [reply]
I have a feeling that I may have screwed things up by accidently choosing a local storage variable to be called subscriptionStatus in order to store the return from the extpay object a couple of days ago. I changed the name to a different name 'subscriberStatus' a day or so ago, but I think it confused the extPay sync, and it has retained a value that I assigned to that local storage variable "Cancelled", and even though I have cleared all local storage values it isn't able to get rid of that value. Is there a way you could reset that user object to test the theory, please?

**Reply by Glench** [reply]
Can you email me the URL of your subscription management page or send me the api key? glen@extensionpay.com

**Reply by FlashInThePan** [reply]
Sorry Glen, which api key/subscription management page do you mean? Stripe, or extpay? I don't think I got an api key for extpay did I? just an extension id

**Reply by Glench** [reply]
Just do extpay.openPaymentPage() in your extension and send me the url of the page that opens.

**Reply by FlashInThePan** [reply]
As I said by email, all is well now.. Thanks for all the advice. Was caused by a typo in an if statement (used single = sign rather than ===, and seems to have been taken as an assignment of cancelled rather than an check for its value). Just thought I would note the resolution here for the forum record.

**Reply by connallykendrick** [reply]
I'm in test mode trying to use my dev email for trial, however it just redirects to Log In once I submit that email.

I've used another email, but it's not showing up under "Free trial users" for me to manage/delete to use again.

**Reply by connallykendrick** [reply]
This still doesn't work for me @Glench - is this feature still implemented? When I load a new version of the extension and input my dev email on the trial sign up page, it redirects to the paid feature login page

**Reply by Glench** [reply]
That's because the email you're using has been used to pay for the extension so it's asking you to log in instead of resume your trial. Test with another email that hasn't paid.

**Reply by ba32107** [reply]
FWIW, now I can indeed try the free trial workflow over and over again, but now I'm not able to test the "real" trial workflow. As in, ExtPay never prevents me from signing up multiple times. I almost filed a bug report now because I thought something wasn't working properly :) 

A proper developer dashboard would be really really useful where I can track the test users and set their status, reset them, delete them, etc both for free trials and for the subscriptions.

**Reply by vinodkumarh** [reply]
In a similar boat. I think I can start the trial after providing my email and refreshing browser (still trying to figure [this](https://github.com/Glench/ExtPay/discussions/220) out).
But not sure if the clock started ticking for trial in test mode. Some way to check that like dev dashboard or something even simpler would be great

**Reply by Glench** [reply]
If you view the changelog on the homepage, you can see that we now let users who have already started a trial "log in" to an existing trial instead of disallowing them completely. They can do this an unlimited amount of times, but their trialStartedAt datetime will always remain the same.

We made this change because it was very confusing for users who had already signed up for a trial, uninstalled the extension, reinstalled, then tried to "log in" (ie resume their existing trial) and received a message that their email had already been used and that they couldn't log in. Since making this change we've received absolutely no complaints whatsoever from users, so we view that as a success! We'll update the docs to explain this a little so it's less confusing.

In development mode, just continue to use your dev email. This is how you test the "real" way of logging in. You can see all trial users in your extension's trial dashboard, shown below:

![Screenshot 2024-07-09 at 10 04 45 AM](https://github.com/Glench/ExtPay/assets/451461/cf8317f2-f2ba-4cc5-b940-9cd4025050ae)

**Reply by ba32107** [reply]
> They can do this an unlimited amount of times, but their trialStartedAt datetime will always remain the same.

Just to confirm, this is not true for my developer email, correct? Because I tried this, and any time I signed up for a trial with my dev email, `trialStartedAt` always reset to the latest value.

**Reply by Glench** [reply]
Exactly. I know it's a bit confusing, but I think it's the best flow that allows devs to test unlimited new trials while in development. If you want to test the trial login function, you can just use a different email. We've updated the README in the next version of ExtPay to make this more explicit.
